### PR TITLE
Sign-in viewer API

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #4227. 

/cc @sebastianbenz , @ericfs for the sign-in protocol preview

### CODE REVIEW

File: extensions/amp-access/0.1/signin.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+ * limitations under the License.
+ */
+
+import {isArray} from '../../../src/types';
+import {isExperimentOn} from '../../../src/experiments';
+import {user} from '../../../src/log';
+
+
+/** @const */
+const TAG = 'amp-access-signin';
```
Comment (erwinmombay) [inline extensions/amp-access/0.1/signin.js]: can we rename to `EXPERIMENT`?

File: extensions/amp-access/0.1/signin.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      if (configSignInServices) {
+        user.assert(isArray(configSignInServices),
+            '"signinServices" must be an array');
+      }
+
+      /** @private @const {boolean} */
+      this.supportsSignInService_ = configSignInServices &&
+          configSignInServices.indexOf(viewerSignInService) != -1;
+
+      /** @private {!Promise<?string>} */
```
Comment (erwinmombay) [inline extensions/amp-access/0.1/signin.js]: `{?Promise<string?>}`

File: extensions/amp-access/0.1/signin.js [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  /**
+   * Whether this protocol has been enabled by the viewer.
+   * @return {boolean}
+   */
+  isEnabled() {
+    return this.isEnabled_;
+  }
+
+  /**
+   */
```
Comment (erwinmombay) [inline extensions/amp-access/0.1/signin.js]: can we remove?

File: extensions/amp-access/0.1/signin.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    this.accessTokenPromise_ = Promise.resolve(accessToken);
+  }
+
+  /**
+   * Processes login dialog's result. And if `#access_grant=` hash parameter
+   * is specified, will send this access grant to the viewer to exchange for
+   * the access token. This method returns promise that will resolve the
+   * exchanged access token.
+   *
+   * If viewer has not opted-in into signin protocol, this method returns
```
Comment (erwinmombay) [inline extensions/amp-access/0.1/signin.js]: nice explanation

File: extensions/amp-access/0.1/signin.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+import {isExperimentOn} from '../../../src/experiments';
+import {user} from '../../../src/log';
+
+
+/** @const */
+const TAG = 'amp-access-signin';
+
+
+/**
+ * This class represents the sign-in protocol, but means of which the source
```
Comment (jridgewell) [inline extensions/amp-access/0.1/signin.js]: s/but/by/

File: extensions/amp-access/0.1/amp-access.js [hunk 6]
```diff
@@ -794,14 +783,19 @@ export class AccessService {
       } else {
         this.loginAnalyticsEvent_(type, 'rejected');
       }
+      const exchangePromise = this.signIn_.postLoginResult(query) ||
+          Promise.resolve();
       if (success || !s) {
```
Comment (erwinmombay) [inline extensions/amp-access/0.1/amp-access.js]: why do we need the extra `!s` check here?

File: extensions/amp-access/0.1/signin.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    /** @private @const {boolean} */
+    this.isEnabled_ =
+        isExperimentOn(this.win, TAG) &&
+        this.viewer_.isEmbedded() &&
+        this.viewer_.getParam('signin') == '1';
+
+    if (this.isEnabled_) {
+
+      /** @private @const {boolean} */
+      this.acceptAccessToken_ = !!configJson['acceptAccessToken'];
```
Comment (jridgewell) [inline extensions/amp-access/0.1/signin.js]: This should be set to `false` if `!isEnabled`

File: extensions/amp-access/0.1/signin.js [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+      const viewerSignInService = this.viewer_.getParam('signinService');
+      const configSignInServices = configJson['signinServices'];
+      if (configSignInServices) {
+        user.assert(isArray(configSignInServices),
+            '"signinServices" must be an array');
+      }
+
+      /** @private @const {boolean} */
+      this.supportsSignInService_ = configSignInServices &&
```
Comment (jridgewell) [inline extensions/amp-access/0.1/signin.js]: Set to `false` if not enabled.

File: extensions/amp-access/0.1/signin.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+          configSignInServices.indexOf(viewerSignInService) != -1;
+
+      /** @private {!Promise<?string>} */
+      this.accessTokenPromise_ = null;
```
Comment (jridgewell) [inline extensions/amp-access/0.1/signin.js]: Should be moved outside the conditional.

File: extensions/amp-access/0.1/signin.js (hunk 7 above)
Comment (erwinmombay) [inline extensions/amp-access/0.1/signin.js]: agree 👍

File: extensions/amp-access/0.1/signin.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: I thought the latest was to call these `TAG`. No? FWIW, it's used as a tag as well in log calls.

File: extensions/amp-access/0.1/signin.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: Done

File: extensions/amp-access/0.1/signin.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: I added docs. But generally, any method should have jsdoc, I believe. Even if empty.

File: extensions/amp-access/0.1/signin.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: Done

File: extensions/amp-access/0.1/amp-access.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: `s` here is the hash string, so it's always a string. In this code it's processed explicitly to show that an empty string is interpreted as `success=yes` for the purposes of re-authorization, but not for other purposes.

File: extensions/amp-access/0.1/signin.js (hunk 7 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: Good catch. Done.

File: extensions/amp-access/0.1/signin.js (hunk 8 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: Done

File: extensions/amp-access/0.1/signin.js (hunk 9 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/signin.js]: Done

**Comment by dvoytenko** [comment]
@erwinmombay @jridgewell Changes are done. PTAL.

File: extensions/amp-access/0.1/amp-access.js (hunk 6 above)
Comment (erwinmombay) [inline extensions/amp-access/0.1/amp-access.js]: i see, so `success=no`, `success=false` has no meaning here?

File: extensions/amp-access/0.1/signin.js (hunk 1 above)
Comment (erwinmombay) [inline extensions/amp-access/0.1/signin.js]: sgtm

File: extensions/amp-access/0.1/amp-access.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: Not quite. `success=no` will not proceed with re-authorization.

**Comment by erwinmombay** [comment]
@dvoytenko LGTM

**Pull request merged**